src/Microsoft.AspNet.Server.Kestrel/Server.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Hosting;
+using Microsoft.AspNet.Hosting.Server;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Features;
+using Microsoft.AspNet.Server.Kestrel.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Server.Kestrel
+{
+    public class Server : IServer
+    {
+        private IApplicationLifetime _applicationLifetime;
+        private ILogger _logger;
+        private Stack<IDisposable> _disposables;
+        private IHttpContextFactory _httpContextFactory;
+
+        public IFeatureCollection Features { get; }
+
+        public Server(IFeatureCollection features, IApplicationLifetime applicationLifetime, ILogger logger, IHttpContextFactory httpContextFactory)
+        {
+            if (features == null)
+            {
+                throw new ArgumentNullException(nameof(features));
+            }
+            if (applicationLifetime == null)
+            {
+                throw new ArgumentNullException(nameof(applicationLifetime));
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            if (httpContextFactory == null)
+            {
+                throw new ArgumentNullException(nameof(httpContextFactory));
+            }
+
+            _applicationLifetime = applicationLifetime;
+            _logger = logger;
+            Features = features;
+            _httpContextFactory = httpContextFactory;
+        }
+
+        public void Start(RequestDelegate requestDelegate)
+        {
+            _disposables = new Stack<IDisposable>();
```

Comment:
A trap here. If the start is invoked twice, the previous list of disposable will be forgotten and cause leak if any.